database-commons/src/main/java/io/cdap/plugin/db/connector/AbstractDBSpecificConnector.java

```diff
@@ -0,0 +1,124 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.db.connector;
+
+import com.google.common.collect.Maps;
+import io.cdap.cdap.api.data.batch.InputFormatProvider;
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.cdap.etl.api.batch.BatchConnector;
+import io.cdap.cdap.etl.api.connector.ConnectorContext;
+import io.cdap.cdap.etl.api.connector.SampleRequest;
+import io.cdap.plugin.common.SourceInputFormatProvider;
+import io.cdap.plugin.common.db.AbstractDBConnector;
+import io.cdap.plugin.common.db.DBConnectorPath;
+import io.cdap.plugin.db.CommonSchemaReader;
+import io.cdap.plugin.db.ConnectionConfigAccessor;
+import io.cdap.plugin.db.SchemaReader;
+import io.cdap.plugin.db.batch.source.DataDrivenETLDBInputFormat;
+import org.apache.hadoop.io.LongWritable;
+import org.apache.hadoop.mapreduce.MRJobConfig;
+import org.apache.hadoop.mapreduce.lib.db.DBConfiguration;
+import org.apache.hadoop.mapreduce.lib.db.DBWritable;
+
+import java.io.IOException;
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+
+/**
+ * An Abstract DB Specific Connector those specific DB connectors can inherits
+ * @param <T> the Record type that specific DB Record Reader may return while sample the data with InputFormat
+ */
+public abstract class AbstractDBSpecificConnector<T extends DBWritable> extends AbstractDBConnector
+  implements BatchConnector<LongWritable, T> {
+
+  protected final AbstractDBSpecificConnectorConfig config;
+
+  protected AbstractDBSpecificConnector(AbstractDBSpecificConnectorConfig config) {
+    super(config);
+    this.config = config;
+  }
+
+  public abstract boolean supportSchema();
+
+  protected abstract Class<? extends DBWritable> getDBRecordType();
+
+  protected SchemaReader getSchemaReader() {
+    return new CommonSchemaReader();
+  }
+
+  @Override
+  protected DBConnectorPath getDBConnectorPath(String path) throws IOException {
+    return DBSpecificPath.of(path, supportSchema());
+  }
+
+  @Override
+  public InputFormatProvider getInputFormatProvider(ConnectorContext context, SampleRequest request)
+    throws IOException {
+    DBSpecificPath path = DBSpecificPath.of(request.getPath(), supportSchema());
+    if (path.getTable() == null) {
+      throw new IllegalArgumentException(
+        String.format("Path %s cannot be sampled. Must have table name in the path.", request.getPath()));
+    }
+
+    ConnectionConfigAccessor connectionConfigAccessor = new ConnectionConfigAccessor();
+    if (config.getUser() == null && config.getPassword() == null) {
+      DBConfiguration
+        .configureDB(connectionConfigAccessor.getConfiguration(), driverClass.getName(),
```

Comment:
nit - we normally split line on parameter

Reply:
done :)